process/block/metablock.go

```diff
@@ -437,14 +439,22 @@ func (mp *metaProcessor) checkProofsForShardData(header *block.MetaBlock) error
 			continue
 		}
 
-		if !mp.proofsPool.HasProof(shardData.ShardID, shardData.HeaderHash) {
-			return fmt.Errorf("%w for header hash %s", process.ErrMissingHeaderProof, hex.EncodeToString(shardData.HeaderHash))
-		}
-
 		if !common.ShouldBlockHavePrevProof(shardHeader.hdr, mp.enableEpochsHandler, common.EquivalentMessagesFlag) {
 			continue
 		}
 
+		if !mp.proofsPool.HasProof(shardData.ShardID, shardData.HeaderHash) {
+			log.Trace("could not find proof for shard data, requesting the next shard header", "current hash", hex.EncodeToString(shardData.HeaderHash))
+			err := mp.requestNextHeaderBlocking(shardHeader.hdr.GetNonce()+1, shardData.ShardID)
+			if err != nil {
+				return err
+			}
+
+			if !mp.proofsPool.HasProof(shardData.ShardID, shardData.HeaderHash) {
+				return fmt.Errorf("%w for header hash %s", process.ErrMissingHeaderProof, hex.EncodeToString(shardData.HeaderHash))
+			}
+		}
```

Comment:
we can extract this check into a common function

Reply:
extracted